### PR TITLE
[jax2tf] Fixes for handling of convolutions with shape_polymorphism and enable_xla=False

### PR DESCRIPTION
**Pull request by gnecula**
Issue: #11402

Due to a typo we were running no tests for convolutions with shape
polymorphism and enable_xla=False.

Added a few more tests from #11402 (Thanks @sdenton4).

The main issue was that in presence of shape polymorphism we cannot
just use `x.shape` for a TF value `x` because it will contain `None`
in the place of unknown dimensions. We must use instead the JAX
abstract values.

This does not fix all issues reported in #11402, there is still the
computation of padding or padding="SAME". Commented out the
corresponding test.

### CODE REVIEW

File: jax/experimental/jax2tf/tests/shape_poly_test.py [hunk 1]
```diff
@@ -1276,18 +1276,55 @@ def _make_harness(group_name: str, name: str,
                   jax.grad(lambda x: jnp.sum(jnp.concatenate([x, x], axis=0))),
                   [RandArg((3, 4, 5), _f32)],
                   poly_axes=[(0, 1)]),
+
+    # Issue #11402 InconclusiveDimensionOperation: Dimension polynomial '-1*t' is not a multiple of '2'
```
Comment (gnecula) [inline jax/experimental/jax2tf/tests/shape_poly_test.py:1280]: This will be fixed in an upcoming PR. This fails with enable_xla=False and True, so it is a high priority.

File: jax/experimental/jax2tf/impl_no_xla.py [hunk 2]
```diff
@@ -97,18 +108,22 @@ def pads_to_padtype(in_shape, window_shape, window_strides, padding) -> str:
...
+    padding = no_pad + padding + no_pad
     # Add empty padding for dummy dimension, too.
-    padding = tf.concat([no_pad, padding, no_pad, no_pad], 0)
+    padding = no_pad + padding + no_pad + no_pad
   else:
-    padding = tf.concat([no_pad, padding, no_pad], 0)
-    in_shape = tf.pad(in_shape, padding)
-  return in_shape
+    padding = no_pad + padding + no_pad
+  x = tf.pad(x, padding)
```
[the remark this answers is not in the document]
Comment (sdenton4) [reply jax/experimental/jax2tf/impl_no_xla.py:122]: It definitely should /not/ be indented. This is likely the source of some shape mismatches I'm seeing. Good catch!

File: jax/experimental/jax2tf/impl_no_xla.py [hunk 3]
```diff
@@ -217,19 +232,27 @@ def _conv_general_dilated(
...
       padding, lhs_dilation, rhs_dilation, is_conv1d)
 
-  lhs, rhs = _transpose_for_tf_conv(lhs, rhs, dimension_numbers)
-
-  in_channels = lhs.shape[-1]
-  *rhs_spatial_shapes, _, rhs_out_channel = rhs.shape
+  lhs, lhs_shape, rhs, rhs_shape = _transpose_for_tf_conv(lhs, lhs_shape,
+                                                          rhs, rhs_shape,
+                                                          dimension_numbers)
+  jax2tf._assert_matching_abstract_shape(lhs, lhs_shape)
```
Comment (sdenton4) [inline jax/experimental/jax2tf/impl_no_xla.py:252]: Looks like you could move these asserts into the _transpose_for_tf_conv function to save some logic in the main branch.

File: jax/experimental/jax2tf/impl_no_xla.py [hunk 4]
```diff
@@ -217,19 +232,27 @@ def _conv_general_dilated(
...
   """Implementation of lax.conv_general_dilated_p using XlaConv."""
+  # In presence of shape polymorphism, lhs.shape and rhs.shape may contain
+  # None. The actual dimension polynomial shapes are in _in_avals.
   del lhs_shape, rhs_shape, precision  # Unused arguments.
-  out_shape = jax2tf._aval_to_tf_shape(_out_aval)
-  _validate_spatial_dimensions(len(lhs.shape) - 2)
-  is_conv1d = len(lhs.shape) - 2 == 1
+  lhs_shape, rhs_shape = _in_avals[0].shape, _in_avals[1].shape
+  jax2tf._assert_matching_abstract_shape(lhs, lhs_shape)
+  jax2tf._assert_matching_abstract_shape(rhs, rhs_shape)
```
Comment (sdenton4) [inline jax/experimental/jax2tf/impl_no_xla.py:240]: Consider updating _validate_spatial_dimensions to consume (lhs, lhs_shape, rhs, rhs_shape) and moving these asserts into it; they certainly with fit the function name, and it'll simplify reading the main function a little bit.

File: jax/experimental/jax2tf/tests/shape_poly_test.py [hunk 5]
```diff
@@ -1117,7 +1117,7 @@ def _make_harness(group_name: str, name: str,
                   check_result=True,
                   skip_jax_run=True,
                   tol=None,
-                  enable_and_diable_xla=False,
+                  enable_and_disable_xla=False,
                   expect_error=(None, None),
                   **params) -> Union[Harness, Sequence[Harness]]:
```
Comment (sdenton4) [inline jax/experimental/jax2tf/tests/shape_poly_test.py:1122]: Can this **params be removed? 

It seems like there's an over-zealous **params somewhere which allowed the typo to pass through undetected... Might be nice to look around and ensure that the 'leaf' functions don't have **params as an argument.

**Review by sdenton4 — APPROVED** [review]
This looks great; massive thanks for prioritizing this work. I assume you'll want Marc to take a look as well, but this LGTM, FWIW.

**Comment by gnecula** [comment]
@rxsang @maxwillzq PTAL

**Comment by marcvanzee** [comment]
Really awesome change @gnecula, thanks a lot!

**Review by marcvanzee — APPROVED** [review]

**Review by rxsang — APPROVED** [review]
